rust-runtime/aws-smithy-http/src/minimum_throughput.rs

```diff
@@ -0,0 +1,123 @@
+use bytes::Buf;
+use http::HeaderMap;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+use std::time::{Duration, Instant};
+
+pin_project_lite::pin_project! {
+    /// A body-wrapper that will ensure that the wrapped body is emitting bytes faster than some
+    /// `minimum_throughput`.
+    pub struct MinimumThroughputBody<InnerBody> {
+        #[pin]
+        inner: InnerBody,
+        // A record of when and how much data was read
+        throughput_logs: Vec<(Instant, u64)>,
+        // The minimum acceptable throughput. If the amount of data per unit of time returned is
+        // less that this, an error will be returned instead.
+        minimum_throughput: (u64, Duration),
+    }
+}
+
+impl<T: http_body::Body> MinimumThroughputBody<T> {
+    pub fn new(body: T, minimum_throughput: (u64, Duration)) -> Self {
+        Self {
+            inner: body,
+            throughput_logs: Vec::new(),
+            minimum_throughput,
+        }
+    }
+}
+
+impl<T> http_body::Body for MinimumThroughputBody<T>
+where
+    T: http_body::Body<Data = bytes::Bytes, Error = Box<dyn std::error::Error + Send + Sync>>,
+{
+    type Data = T::Data;
+    type Error = T::Error;
+
+    fn poll_data(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Option<Result<Self::Data, Self::Error>>> {
+        let this = self.project();
+
+        let poll_res = this.inner.poll_data(cx);
+
+        if let Poll::Ready(Some(Ok(ref data))) = poll_res {
+            this.throughput_logs
+                .push((Instant::now(), data.remaining() as u64));
+        };
+
+        let mut logs = this.throughput_logs.iter();
+        if let Some((first_instant, first_bytes)) = logs.next() {
+            let time_elapsed_since_first_poll = first_instant.elapsed();
+            let mut total_bytes_read = *first_bytes;
+
+            while let Some((_, bytes_read)) = logs.next() {
+                total_bytes_read += bytes_read;
+            }
+
+            let minimum_bytes_per_second =
+                this.minimum_throughput.0 as f64 / this.minimum_throughput.1.as_secs_f64();
+            let actual_bytes_per_second =
+                total_bytes_read as f64 / time_elapsed_since_first_poll.as_secs_f64();
```

Comment:
I think to be accurate for longer downloads, and to avoid consuming excess memory, you'll need to have a sliding window over the `throughput_logs`. It's probably as simple as popping the front of the vec until the first entry is no longer older than the configured duration (before doing this calculation).

Additionally, we probably shouldn't do this comparison at all unless the configured duration has elapsed already. Otherwise, there's a chance that a slow-to-start download would always result in failure since it would only have one data point.